notebooks/autodiff_cookbook.ipynb

```diff
@@ -0,0 +1,1062 @@
+      ]
+    },
+    {
+      "metadata": {
+        "id": "1tFcRuEzkGRR",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "from jax import vjp\n",
+        "\n",
+        "# Isolate the function from the weight matrix to the predictions\n",
+        "f = lambda W: predict(W, b, inputs)\n",
+        "\n",
+        "y, vjp_fun = vjp(f, W)\n",
+        "\n",
+        "key, subkey = random.split(key)\n",
+        "u = random.normal(subkey, y.shape)\n",
+        "\n",
+        "# Pull back the covector `u` along `f` evaluated at `W`\n",
+        "v = vjp_fun(u)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "oVOZexCEkvv3",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "In terms of Haskell-like type signatures, we could write\n",
+        "\n",
+        "```haskell\n",
+        "vjp :: (a -> b) -> a -> (b, CT b -> CT a)\n",
+        "```\n",
+        "\n",
+        "where we use `CT a` to denote the type for the cotangent space for `a`. In words, `vjp` takes as arguments a function of type `a -> b` and a point of type `a`, and gives back a pair consisting of a value of type `b` and a linear map of type `CT b -> CT a`.\n",
+        "\n",
+        "This is great because it lets us build Jacobian matrices one row at a time, and the FLOP cost for evaluating $(x, v) \\mapsto (f(x), v^\\mathsf{T} \\partial f(x))$ is only about twice the cost of evaluating $f$. In particular, if we want the gradient of a function $f : \\mathbb{R}^n \\to \\mathbb{R}$, we can do it in just one call. That's how `grad` is efficient for gradient-based optimization, even for objectives like neural network training loss functions on millions or billions of parameters.\n",
+        "\n",
+        "There's a cost, though: though the FLOPs are friendly, memory scales with the depth of the computation. Also, the implementation is traditionally more complex than that of forward-mode, though JAX has some tricks up its sleeve (that's a story for a future notebook!).\n",
+        "\n",
+        "For more on how reverse-mode works, see [this tutorial video from the Deep Learning Summer School in 2017](http://videolectures.net/deeplearning2017_johnson_automatic_differentiation/)."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "xtqSUJgzwQXO",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "## Composing VJPs, JVPs, and `vmap`"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "PSL1TciM6vvI",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "### Jacobian-Matrix and Matrix-Jacobian products\n",
+        "\n",
+        "Oh, that's interesting. No more loops?"
```

Comment:
This needs more text to fit with the style of the previous text. Why is it interesting? Complete sentences please!

Reply:
If nothing else, use the opportunity to pitch `vmap`.

Reply:
Added words!